crates/api/src/event/event_adapter.rs

```diff
@@ -0,0 +1,107 @@
+use std::sync::{mpsc, Arc, Mutex};
+
+use dces::entity::Entity;
+
+use crate::{event::*, shell::WindowRequest};
+
+/// The `EventAdapter` provides a thread save way to push events to the widget tree of a window.
+///
+/// # Example
+///
+/// ```rust
+/// impl State for MayState {
+///     fn update(&mut self, _: &mut Registry, ctx: &mut Context) {
+///         let event_adapter = ctx.event_adapter();
+///         let entity = ctx.entity;
+///         
+///         let _ = std::thread::spawn(move|| {
+///             loop {
+///                 // pushes all 10 milliseconds an activate event on the entity of the state
+///                 event_adapter.push_event(entity, ActivateEvent(entity));
+///                 std::thread::sleep(std::time::Duration::from_millis(10));
+///             }
+///         });
+///     }    
+/// }
+/// ```
+#[derive(Clone, Default, Debug)]
+pub struct EventAdapter {
+    event_queue: Arc<Mutex<EventQueue>>,
+    window_sender: Option<mpsc::Sender<WindowRequest>>,
+}
+
+impl EventAdapter {
+    /// Creates a new event adapter.
+    pub fn new(window_sender: mpsc::Sender<WindowRequest>) -> Self {
+        EventAdapter {
+            event_queue: Arc::new(Mutex::new(EventQueue::new())),
+            window_sender: Some(window_sender),
+        }
+    }
+
+    /// Push an event to the tree starting by the given entity. The event bubbles through the tree until it is handled.
+    pub fn push_event<E: Event + Send>(&self, entity: Entity, event: E) {
+        self.event_queue
+            .lock()
+            .expect("EventAdapter::push_event: Cannot lock event queue.")
+            .register_event(event, entity);
+
+        self.redraw();
+    }
+
+    /// Pushes an event that is direct send to the given entity (widget). I occurs only by the given entity and will not bubble through the tree.
```

Comment:
I think this should be
"Pushes an event that is directly sent to the given entity (widget). It occurs only by the given entity and will not bubble through the tree."
but a native speaker should review this too.